internal/controllers/operator_controller.go

```diff
@@ -459,3 +478,27 @@ func operatorRequestsForCatalog(ctx context.Context, c client.Reader, logger log
 		return requests
 	}
 }
+
+// Complete fills in the annotations from the information received from
+// bundleEntities.
+func (bdm *bundleDeploymentMetadata) CompleteBundleDeploymentMetadata(entity *entity.BundleEntity) error {
+	var errs []error
+
+	channel, err := entity.ChannelName()
+	if err != nil || channel == "" {
+		errs = append(errs, fmt.Errorf("unable to find the channel name from resolved entity: %v", err))
+		channel = "unknown"
+	}
+	bdm.channel = channel
+
+	var version string
+	semverVer, err := entity.Version()
+	version = semverVer.String()
+	if err != nil || version == "" {
+		errs = append(errs, fmt.Errorf("unable to find the version of operator to be installed from resolved entity: %v", err))
+		version = "unknown"
+	}
+	bdm.version = version
```

Comment:
I'm honestly not even sure an annotation-based approach is the right one. The thing that uniquely identifies the thing that we have installed is the bundle image SHA which is already in the BD.

Could the resolver not:
1. Find the current image sha of the BD
2. Locate all entities whose image == the current image sha
3. Find the entities that say they can upgrade from the entities we found in (2)
4. Then build a constraint such that the resolved entity must come from set 2 (currently installed) or set 3 (successors of currently installed)

Reply:
We would need to think about the ramifications of various scenarios of the size of set (2):
- Size 0: we don't know what's installed, therefore can't upgrade
- Size 1: we know what's installed, therefore we can upgrade or stick to what we're on
- Size >1: there are multiple entities that specify the image ref we're looking for, but we don't know which one we derived the BD from. There are other aspects of the entity (e.g. `olm.bundle.mediatype`) that impact the BD, so we would NOT want to end up selecting an entity that resulted in a side-grade to a different BD spec (despite the same bundle image)

Reply:
That could be possible, but it could introduce ambiguity. I.e. you could have two bundles pointing to the same image sha, because of reasons. That's why I'd argue it would be better to have an exact 1:1 between the BD and the underlying bundle/entity. Maybe the sane alternative is to rather than break it down along the parameters of the bundle (package, version, channel), just use the entities unique ID. Then we're safe no matter what.

Reply:
@perdasilva Agree to what you have mentioned. But I still have questions on the implementation part of it to make it possible.

> Locate all entities whose image == the current image sha

How do we work back on locating the entities when they (or neither the solution set) is persisted anywhere during the reconciliation. We would still need to query for the installed BDs on the cluster and provide the necessary info to re-calculate the entities? The installed bundle can be from a catsrc which is not even available on cluster.

Reply:
Straw man: the unique ID is the image SHA.
- The image SHA is content-addressable. If anything about the bundle changes, the SHA changes.
- The bundle image contains ALL of the metadata used for resolution and there is no way to change that metadata without changing the SHA, so if we see an image SHA twice in the catalog, we also know for sure that its resolution parameters are identical.

Reply:
> We would still need to query for the installed BDs on the cluster and provide the necessary info to re-calculate the entities? 

I think that's what I'm suggesting, yes.

> The installed bundle can be from a catsrc which is not even available on cluster.

And that's okay. There are two potential scenarios there:
1. Nothing currently on the cluster says it upgrades from it, so we just don't upgrade it
2. Some other catalog source has an upgrade edge that specifies the currently installed bundle SHA as the `from` version, which case we can upgrade to the `to` version.

Reply:
This is definitely debatable though. I think I could also make a pretty good argument for the unique ID being `packageName` + `bundleVersion` + `bundleRelease`

And maybe if there are two bundles with that same 3 tuple available on the catalog, we fail resolution and ask the user to do provide a `includes` or `excludes` configuration on either an entire catalog source or on a catsrc + package tuple.

Reply:
> The entity should have an unique id, which I guess is what we should store in the annotations. Then, you could just use the entitysource.Get(id) method to get the entity back on the next resolution. Would that work?

If we do that, I think we need a spec for entity ID, and we need to consider what happens if we can't find that entity ID we stored in the BD annotation during the next resolution. For example, right now the catalogd entity source includes `catalog.metadata.name` in the entity ID. If someone changes the name of a catalog, we probably don't want resolution to stop working for any entities that were installed from the previous name.